server/src/main/java/io/crate/replication/logical/ShardReplicationChangesTracker.java

```diff
@@ -0,0 +1,220 @@
+                     shardId, seqNoAlreadyRequested.get(), observedSeqNoAtLeader.get());
+
+        // Wait till we have batch to fetch. Note that if seqNoAlreadyRequested is equal to observedSeqNoAtLeader,
+        // we still should be sending one more request to fetch which will just do a poll and eventually timeout
+        // if no new operations are there on the leader (configured via TransportGetChangesAction.WAIT_FOR_NEW_OPS_TIMEOUT)
+        if (seqNoAlreadyRequested.get() > observedSeqNoAtLeader.get() && missingBatches.isEmpty()) {
+            cancellable = threadPool.schedule(
+                () -> requestBatchToFetch(consumer),
+                replicationSettings.pollDelay(),
+                ThreadPool.Names.LOGICAL_REPLICATION
+            );
+            return;
+        }
+
+        // missing batch takes higher priority.
+        if (missingBatches.isEmpty() == false) {
+            var missingBatch = missingBatches.remove(0);
+            LOGGER.debug("[{}] Fetching missing batch {}-{}", shardId, missingBatch.v1(), missingBatch.v2());
+            consumer.accept(missingBatch);
+        } else {
+            // return the next batch to fetch and update seqNoAlreadyRequested.
+            var batchSize = replicationSettings.batchSize();
+            var fromSeq = seqNoAlreadyRequested.getAndAdd(batchSize) + 1;
+            var toSeq = fromSeq + batchSize - 1;
+            LOGGER.debug("[{}] Fetching the batch {}-{}", shardId, fromSeq, toSeq);
+            consumer.accept(new Tuple<>(fromSeq, toSeq));
+        }
+    }
+
+    private ShardChangesAction.Response getChanges(Long fromSeqNo, Long toSeqNo) {
+        var remoteClient = remoteClientResolver.apply(shardId.getIndexName());
+        if (remoteClient == null) {
+            throw new RuntimeException("Could not resolve a remote client for index=" + shardId.getIndexName());
+        }
+        var request = new ShardChangesAction.Request(shardId, fromSeqNo, toSeqNo);
+        return remoteClient.execute(ShardChangesAction.INSTANCE, request).actionGet(WAIT_FOR_NEW_OPS_TIMEOUT);
+    }
+
+
+    /**
+     * Ensures that we've successfully fetched a particular range of operations.
+     * In case of any failure(or we didn't get complete batch), we make sure that we're fetching the
+     * missing operations in the next batch.
+     */
+    private void updateBatchFetched(boolean success,
+                                    long fromSeqNoRequested,
+                                    long toSeqNoRequested,
+                                    long toSeqNoReceived,
+                                    long seqNoAtLeader) {
+        if (success) {
+            // we shouldn't ever be getting more operations than requested.
+            assert toSeqNoRequested >= toSeqNoReceived :
+                Thread.currentThread().getName() + " Got more operations in the batch than requested";
+            LOGGER.debug("Updating the batch fetched. {}-{}/{}, seqNoAtLeader:{}",
```

Comment:
How often is this code path executed? Would it make sense to check if debug level is enabled?